Kernel/FileSystem/Ext2FS/Inode.cpp

```diff
@@ -417,10 +427,38 @@ InodeMetadata Ext2FSInode::metadata() const
     metadata.uid = inode_uid(m_raw_inode);
     metadata.gid = inode_gid(m_raw_inode);
     metadata.link_count = m_raw_inode.i_links_count;
-    metadata.atime = UnixDateTime::from_seconds_since_epoch(static_cast<i32>(m_raw_inode.i_atime));
-    metadata.ctime = UnixDateTime::from_seconds_since_epoch(static_cast<i32>(m_raw_inode.i_ctime));
-    metadata.mtime = UnixDateTime::from_seconds_since_epoch(static_cast<i32>(m_raw_inode.i_mtime));
-    metadata.dtime = UnixDateTime::from_seconds_since_epoch(static_cast<i32>(m_raw_inode.i_dtime));
+
+    // NOTE: All the *_extra fields have to be bounds-checked in case we have oddly-sized inodes.
+    // This is simply a correctness measure, since an OOB read wouldn't happen anyway due to the
+    // fact that we always store the raw inode as an ext2_inode_large.
+    if (is_within_inode_bounds(bit_cast<FlatPtr>(&m_raw_inode), bit_cast<FlatPtr>(&m_raw_inode.i_atime_extra), sizeof(m_raw_inode.i_atime_extra))) {
+        time_t seconds = decode_seconds_with_extra(m_raw_inode.i_atime, m_raw_inode.i_atime_extra);
+        u32 nanoseconds = decode_nanoseconds_from_extra(m_raw_inode.i_atime_extra);
+        metadata.atime = UnixDateTime::from_unix_timespec({ .tv_sec = seconds, .tv_nsec = nanoseconds });
+    } else {
+        metadata.atime = UnixDateTime::from_seconds_since_epoch(static_cast<i32>(m_raw_inode.i_atime));
+    }
+
+    // NOTE: There's no i_dtime_extra, so we use i_ctime_extra to approximate the right epoch for metadata.dtime (like e2fsprogs.)
```

Comment:
e2fsprogs is GPL I think, so we shouldn't have comments that sound like we looked at its source. Maybe this is easy to see from its external behavior?